File: dynamicprogramming/fibonacci.js
let calc=0
function dpfibonacci(n){
    let cache={0:0,1:1,2:1}
    return function fib(n){
        if(n in cache){
            return cache[n]
        }
        else{
            cache[n]=fib(n-1)+fib(n-2)
            return cache[n]
        }
    }
}
const mem=dpfibonacci()
console.log(mem(1000))

const fib=(n)=>{
    const res=Math.floor((((1+Math.sqrt(5))**n)-((1-Math.sqrt(5)))**n)/(2**n*Math.sqrt(5)))
    return res
}
console.log(fib(100)) //time complexity o(1)
                    //space cpmplexity o(1)
                    //however it's not a readable solution

File: dynamicprogramming/fibonacci.py
from functools import lru_cache
from math import sqrt
#using lru cache decorator
# @lru_cache(maxsize=32)
# def fib(n):
#     if n<2:
#         return n
#     else:
#         return fib(n-1)+fib(n-2)
# print(fib(1000))


# def dpfib(n):
#     cache={0:1,1:1,2:1}
#     def wrapfunc(n):
#         if n in cache or n==3:
#             return cache[n]
#         else:
#             cache[n]=wrapfunc(n-1)+wrapfunc(n-2)
#             return cache[n]
#     return wrapfunc(n)
# mem=dpfib(1000)
# print(mem)

def fib(n):
    res=(((1+sqrt(5))**n)-((1-sqrt(5)))**n)//(2**n*sqrt(5)) 
    return res
print(fib(100))